File: sam_g2p.rb
=begin
Iterate through a CSV (produced by remap.py) where each row contains
 the contents of a SAM file.
Sequences have already been restricted to HIV env V3 if the reads were
 mapped to the 'V3LOOP' region.
Do not attempt to clean sequence based on CIGAR string.
Apply Conan's implementation of the geno2pheno (g2p) coreceptor
 tropism prediction algorithm to each entry.  Output tuples of header,
 g2p score, FPR, and g2p-aligned protein sequence.

Call:
ruby sam_to_g2p.rb <INPUT SAM csv> <OUTPUT scored csv>

Dependencies:
BioRuby
pssm_lib.rb
g2p.matrix
CSV [Ruby module]
=end

require 'csv'
require './pssm_lib'
require 'bio'
require 'ostruct'

QMIN = 20   # minimum base quality within insertions
QCUT = 10   # minimum base quality to not be censored
QDELTA = 5

def parse_options()
  options = OpenStruct.new
  options.remap_csv, options.g2p_csv = ARGV
  raise "Usage: #{__FILE__} remap_csv g2p_csv" unless options.g2p_csv
  
  options
end

# Applies a cigar string to recreate a read, then clips the read.
#
# * <tt>:cigar</tt> - a string in the CIGAR format, describing the relationship
#   between the read sequence and the consensus sequence
# * <tt>:seq</tt> - the sequence that was read
# * <tt>:qual</tt> - quality codes for each base in the read
# * <tt>:pos</tt> - first position of the read, given in consensus coordinates
# * <tt>:clip_from</tt> - first position to include after clipping, given in
#   consensus coordinates
# * <tt>:clip_to</tt> - last position to include after clipping, given in
#   consensus coordinates, nil means no clipping at the end
#
# Returns the new sequence and the new quality string
#
#   apply_cigar_and_clip("12M", "AAACAACCACCC", "AAAAAAAAAAAA", 0, 3, 9)
#   # => "CAACCA", "AAAAAA"
def apply_cigar_and_clip(cigar, seq, qual, pos=0, clip_from=0, clip_to=nil)
  clip_to = clip_to || -1
  newseq = '-' * pos
  newqual = '!' * pos
  is_valid =        /^((\d+)([MIDNSHPX=]))*$/.match(cigar)
  tokens = cigar.scan(/(\d+)([MIDNSHPX=])/)
  raise ArgumentError.new("Invalid CIGAR string: '#{cigar}'.") unless is_valid
  left = 0
  tokens.each do |token|
      length, operation = token
      length = length.to_i
      right = left + length - 1
      if operation == 'M'
          newseq += seq[left..right]
          newqual += qual[left..right]
          left += length
      elsif operation == 'D'
          newseq += '-'*length
          newqual += ' '*length
      elsif operation == 'I'
          its_quals = qual[left..right].unpack('C*') # array of ASCII codes
          if its_quals.map{|asc| asc-33 >= QMIN}.all?
              # accept only high quality insertions
              newseq += seq[left..right]
              newqual += qual[left..right]
          end
          left += length
      elsif operation == 'S'
          left += length
      else
          raise ArgumentError.new("Unsupported CIGAR token: '#{token}'.")
      end
      if left > seq.length
        message = "CIGAR string '#{cigar}' is too long for sequence '#{seq}'."
        raise ArgumentError.new(message)
      end
  end
  if left < seq.length
    message = "CIGAR string '#{cigar}' is too short for sequence '#{seq}'."
    raise ArgumentError.new(message)
  end
  newseq = newseq[clip_from..clip_to]
  newqual = newqual[clip_from..clip_to]
  return newseq, newqual
end

def merge_pairs(seq1, seq2, qual1, qual2)
    mseq = ''
    if seq2.length < seq1.length # make sure seq2 is longer
      seq1, seq2 = seq2, seq1 
      qual1, qual2 = qual2, qual1
    end
    i = 0
    qual1_ints = qual1.unpack('C*')
    qual2_ints = qual2.unpack('C*')
    qual1_ints.map! {|x| x - 33}
    qual2_ints.map! {|x| x - 33}
    
    seq2.each_byte { |b2|
        c2 = b2.chr
        q2 = qual2_ints[i]
        if i < seq1.length
            c1 = seq1[i,1]
            q1 = qual1_ints[i]
            if c1 == '-' and c2 == '-'
                mseq += '-'
            elsif c1 == c2
                if q1 > QCUT or q2 > QCUT
                    mseq += c1
                else
                    mseq += 'N'
                end
            else
                if (q2 - q1).abs >= QDELTA
                    if q1 > [q2, QCUT].max
                        mseq += c1
                    elsif q2 > QCUT
                        mseq += c2
                    else
                        mseq += 'N'
                    end
                else
                    mseq += 'N'
                end
            end
        else
            # past the end of read 1
            if c2 == '-' and q2 == 0
                mseq += 'n' # interval between reads
            else
                if q2 > QCUT
                    mseq += c2
                else
                    mseq += 'N'
                end
            end
        end
        i += 1 # increment at end of loop
    }
    return mseq
end

def main()
  options = parse_options()
  ## convert file contents into hash to collect identical variants
  
  pairs = Hash.new  # cache read for pairing
  merged = Hash.new # tabulate merged sequence variants
  sample_name = ''
  
  CSV.foreach(options.remap_csv) do |row|
      sample_name, qname, flag, rname, pos, mapq, cigar, rnext, pnext, tlen, seq, qual = row
      if rname != 'HIV1B-env-seed'
          # uninteresting region or the header row
          next
      end
      seq1, qual1 = apply_cigar_and_clip(cigar, seq, qual)
      pos1 = pos.to_i-1
      seq2 = '-'*pos1 + seq1 # pad the sequence
      qual2 = '!'*pos1 + qual1
  
      pair = pairs[qname]
      if pair == nil
        # cache this read
        pairs[qname] = {'seq'=>seq2, 'qual'=>qual2}
      else
        # merge reads
        pairs.delete(qname)
        seq1 = pair['seq'] # retrieve from cache
        qual1 = pair['qual']
      
        mseq = merge_pairs(seq1, seq2, qual1, qual2)
      
        count = merged.fetch(mseq, 0)
        merged[mseq] = count + 1
      end
  end  
  
  sorted = merged.sort_by {|k,v| v}.reverse  # sort by count in descending order
  
  ## apply g2p algorithm to merged reads
  f = File.open(options.g2p_csv, mode='w') # write-only
  f.write("sample,rank,count,g2p,fpr,aligned,error\n")  # CSV header
  
  rank = 0
  sorted.each do |s, count|
      rank += 1
      prefix = "#{sample_name},#{rank},#{count}"
      seq = s.delete '-'
      if (seq.count 'N') > (0.5*seq.length)
          # if more than 50% of sequence is garbage
          f.write("#{prefix},,,,low quality\n")
          next
      end
      
      if s.length % 3 != 0
          f.write("#{prefix},,,,notdiv3\n")
          next
      end
    
      if seq.length == 0
        f.write("#{prefix},,,,zerolength\n")
        next
      end
    
      dna = Bio::Sequence.auto(seq)
      prot = dna.translate
  
      # sanity check 1 - bounded by cysteines
      if !prot.match(/^C/) || !prot.match(/C$/)
          f.write("#{prefix},,,#{prot},cysteines\n")
          next
      end
  
      # sanity check 2 - no ambiguous codons
      if prot.count('X') > 0
          f.write("#{prefix},,,#{prot},ambiguous\n")
          next
      end
  
      # sanity check 3 - no stop codons
      if prot.count('*') > 0
          f.write("#{prefix},,,#{prot},stop codons\n")
          next
      end
  
      # sanity check 4 - V3 length in range 32-40 inclusive
      if prot.length < 32 || prot.length > 40
          f.write("#{prefix},,,#{prot},length\n")
          next
      end
  
      score, aligned = run_g2p(seq, $std_v3_g2p, load_matrix('g2p'))
      begin
          aligned2 = aligned.flatten * ""
      rescue
          # sequence failed to align
          f.write("#{prefix},#{score},,,failed to align\n")
          next
      end
      fpr = g2p_to_fpr(score)
      f.write("#{prefix},#{score},#{fpr},#{aligned2},\n")
  end
  
  f.close()
end

if __FILE__ == $0
  main()
end

File: micall/g2p/sam_g2p.py
#! /usr/bin/env python

import sys
import re
import argparse
from csv import DictReader
from micall.core.sam2aln import merge_pairs, cigar_re
from micall.utils.translation import translate

# screen for in-frame deletions
pat = re.compile('([A-Z])(---)+([A-Z])')

QMIN = 20   # minimum base quality within insertions
QCUT = 10   # minimum base quality to not be censored
QDELTA = 5

def parse_args():
    parser = argparse.ArgumentParser(description='Calculate g2p scores from amino acid sequences.')

    parser.add_argument('remap_csv', type=argparse.FileType('rU'),
                        help='<input> CSV containing remap output (modified SAM)')
    parser.add_argument('nuc_csv', type=argparse.FileType('rU'),
                        help='<input> CSV containing nucleotide frequency output from aln2counts.py')
    parser.add_argument('g2p_csv', type=argparse.FileType('w'),
                        help='<output> CSV containing g2p predictions.')
    return parser.parse_args()


class RegionTracker:
    def __init__(self, tracked_region):
        self.tracked_region = tracked_region
        self.ranges = {}

    def add_nuc(self, seed, region, query_pos):
        """
         # Add a nucleotide position to the tracker.
        :param seed: name of the seed region
        :param region: name of the coordinate region
        :param query_pos: query position in the consensus coordinates
        :return: unused
        """
        if region != self.tracked_region:
            return

        if seed in self.ranges:
            range = self.ranges[seed]
            if range[1] < query_pos:
                range[1] = query_pos
            elif query_pos < range[0]:
                range[0] = query_pos
        else:
            self.ranges.update({seed: [query_pos, query_pos]})

    def get_range(self, seed):
        """
        Get the minimum and maximum query positions that were seen for a seed.
        :param seed: name of the seed region
        :return: array of two integers
        """
        return self.ranges.get(seed, [None, None])


def apply_cigar_and_clip (cigar, seq, qual, pos=0, clip_from=0, clip_to=None):
    """
    Use CIGAR string (Compact Idiosyncratic Gapped Alignment Report) in SAM data
    to apply soft clips, insertions, and deletions to the read sequence.
    Any insertions relative to the sample consensus sequence are discarded to
    enforce a strict pairwise alignment, and returned separately in a
    dict object.
    """
    newseq = '-' * int(pos)  # pad on left
    newqual = '!' * int(pos)
    tokens = cigar_re.findall(cigar)
    if len(tokens) == 0:
        return None, None, None
    left = 0
    for token in tokens:
        length = int(token[:-1])
        # Matching sequence: carry it over
        if token[-1] == 'M':
            newseq += seq[left:(left+length)]
            newqual += qual[left:(left+length)]
            left += length
        # Deletion relative to reference: pad with gaps
        elif token[-1] == 'D':
            newseq += '-'*length
            newqual += ' '*length  # Assign fake placeholder score (Q=-1)
        # Insertion relative to reference
        elif token[-1] == 'I':
            its_quals = qual[left:(left+length)]
            if all([(ord(q)-33) >= QMIN for q in its_quals]):
                # accept only high quality insertions relative to sample consensus
                newseq += seq[left:(left+length)]
                newqual += its_quals
                if clip_to:
                    clip_to += length  # accommodate insertion
            left += length
        # Soft clipping leaves the sequence in the SAM - so we should skip it
        elif token[-1] == 'S':
            left += length
        else:
            print "Unable to handle CIGAR token: {} - quitting".format(token)
            sys.exit()

    return newseq[clip_from:clip_to], newqual[clip_from:clip_to]


def sam_g2p(pssm, remap_csv, nuc_csv, g2p_csv):
    pairs = {}  # cache read for pairing
    merged = {}  # tabular merged sequence variants
    tracker = RegionTracker('V3LOOP')

    # look up clipping region for each read
    reader = DictReader(nuc_csv)
    for row in reader:
        if row['query.nuc.pos'] == '':
            # skip deletions in query relative to reference
            continue
        tracker.add_nuc(row['seed'], row['region'], int(row['query.nuc.pos'])-1)

    # parse contents of remap CSV output
    reader = DictReader(remap_csv)
    for row in reader:
        clip_from, clip_to = tracker.get_range(row['rname'])
        if clip_from is None:
            # uninteresting region
            continue

        seq2, qual2 = apply_cigar_and_clip(row['cigar'], row['seq'], row['qual'], int(row['pos'])-1, clip_from, clip_to+1)

        mate = pairs.get(row['qname'], None)
        if mate:
            seq1 = mate['seq']
            qual1 = mate['qual']

            # merge pairs only if they are the same length
            if len(seq1) == len(seq2):
                mseq = merge_pairs(seq1, seq2, qual1, qual2)
            else:
                # implies an insertion not covered in one read mate
                mseq = seq1 if len(seq1) > len(seq2) else seq2

            if mseq not in merged:
                merged.update({mseq: 0})
            merged[mseq] += 1

        else:
            pairs.update({row['qname']: {'seq': seq2, 'qual': qual2}})


    sorted = [(v,k) for k, v in merged.iteritems()]
    sorted.sort(reverse=True)

    # apply g2p algorithm to merged reads
    with g2p_csv as f:
        f.write('rank,count,g2p,fpr,aligned,error\n')  # CSV header
        rank = 0
        for count, s in sorted:
            # remove in-frame deletions
            seq = re.sub(pat, r'\g<1>\g<3>', s)

            rank += 1
            prefix = '%d,%d' % (rank, count)
            seqlen = len(seq)
            if seq.upper().count('N') > (0.5*seqlen):
                # if more than 50% of the sequence is garbage
                f.write('%s,,,,low quality\n' % prefix)
                continue

            if seqlen == 0:
                f.write('%s,,,,zerolength\n' % prefix)
                continue

            prot = translate(seq, 0, ambig_char='X')

            # sanity check 1 - bounded by cysteines
            if not prot.startswith('C') or not prot.endswith('C'):
                f.write('%s,,,%s,cysteines\n' % (prefix, prot))
                continue

            # sanity check 2 - too many ambiguous codons
            if prot.count('X') > 2:
                f.write('%s,,,%s,>2ambiguous\n' % (prefix, prot))
                continue

            # sanity check 3 - no stop codons
            if prot.count('*') > 0:
                f.write('%s,,,%s,stop codons\n' % (prefix, prot))
                continue

            # sanity check 4 - V3 length in range 32-40 inclusive
            if len(prot) < 32 or len(prot) > 40:
                f.write('%s,,,%s,length\n' % (prefix, prot))
                continue

            score, aligned = pssm.run_g2p(seq)

            try:
                aligned2 = ''.join([aa_list[0] if len(aa_list) == 1 else '[%s]'%''.join(aa_list)
                                    for aa_list in aligned])
            except:
                # sequence failed to align
                f.write('%s,%s,,,failed to align\n' % (prefix, score))
                continue

            fpr = pssm.g2p_to_fpr(score)
            f.write(','.join(map(str, [prefix, score, fpr, aligned2, 'ambiguous' if '[' in aligned2 else '']))+'\n')

        f.close()


def main():
    args = parse_args()
    from micall.g2p.pssm_lib import Pssm
    pssm = Pssm()
    sam_g2p(pssm=pssm, remap_csv=args.remap_csv, nuc_csv=args.nuc_csv, g2p_csv=args.g2p_csv)

if __name__ == '__main__':
    # note, must be called from project root if executing directly
    # i.e., python micall/g2p/sam_g2p.py -h
    main()
